proxy/src/main/java/com/velocitypowered/proxy/scheduler/VelocityScheduler.java

```diff
@@ -56,273 +56,325 @@
  * {@link ScheduledExecutorService} and a dynamically-sized {@link ExecutorService}.
  * Many plugins are accustomed to the Bukkit Scheduler model, although it is not relevant
  * in a proxy context.
+ * 
+ * <p>This implementation has been optimized for high performance and thread safety
+ * using concurrent collections and atomic operations.</p>
  */
 public class VelocityScheduler implements Scheduler {
 
-  private final PluginManager pluginManager;
-  private final ScheduledExecutorService timerExecutionService;
-  private final Multimap<Object, ScheduledTask> tasksByPlugin = Multimaps.synchronizedMultimap(
-      Multimaps.newSetMultimap(new IdentityHashMap<>(), HashSet::new));
-
-  /**
-   * Initalizes the scheduler.
-   *
-   * @param pluginManager the Velocity plugin manager
-   */
-  public VelocityScheduler(PluginManager pluginManager) {
-    this.pluginManager = pluginManager;
-    this.timerExecutionService = Executors
-        .newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setDaemon(true)
-            .setNameFormat("Velocity Task Scheduler Timer").build());
-  }
-
-  @Override
-  public TaskBuilder buildTask(Object plugin, Runnable runnable) {
-    checkNotNull(plugin, "plugin");
-    checkNotNull(runnable, "runnable");
-    final Optional<PluginContainer> container = pluginManager.fromInstance(plugin);
-    checkArgument(container.isPresent(), "plugin is not registered");
-    return new TaskBuilderImpl(container.get(), runnable);
-  }
-
-  @Override
-  public TaskBuilder buildTask(Object plugin, Consumer<ScheduledTask> consumer) {
-    checkNotNull(plugin, "plugin");
-    checkNotNull(consumer, "consumer");
-    final Optional<PluginContainer> container = pluginManager.fromInstance(plugin);
-    checkArgument(container.isPresent(), "plugin is not registered");
-    return new TaskBuilderImpl(container.get(), consumer);
-  }
-
-  @Override
-  public @NonNull Collection<ScheduledTask> tasksByPlugin(@NonNull Object plugin) {
-    checkNotNull(plugin, "plugin");
-    checkArgument(pluginManager.fromInstance(plugin).isPresent(), "plugin is not registered");
-    final Collection<ScheduledTask> tasks = tasksByPlugin.get(plugin);
-    synchronized (tasksByPlugin) {
-      return Set.copyOf(tasks);
+    private final PluginManager pluginManager;
+    private final ScheduledExecutorService timerExecutionService;
+    
+    /**
+     * Tracks tasks by their plugin instance using concurrent collections.
+     * Uses synchronized IdentityHashSet per plugin for efficient task management.
+     */
+    private final ConcurrentHashMap<Object, Set<ScheduledTask>> tasksByPlugin = new ConcurrentHashMap<>();
```

Comment:
Just as an example of the slop here: this is not only the wrong code style, this is also going to silently do the wrong thing! `ConcurrentHashMap` does not use the identity hash code of the object.

Reply:
I'm perfectly aware, that's why I called it slop.